### PR TITLE
Clean up / document metrics monitor fields

### PR DESCRIPTION
**Pull request by faec**
Document the semantics of many metrics monitoring variables, and rename some metrics APIs to more clearly indicate their function. As a side effect, fix several metrics reporting / publishing bugs in the Elasticsearch output, including https://github.com/elastic/beats/issues/39146.

Add the `output.events.dead_letter` metric to distinguish events that were ingested to the dead letter index after a fatal error (previously these events were just reported as "acked").

This could have been a shorter fix, but it was hard to properly test since the metrics were changed from two separate functions with a lot of special cases. I ended up reorganizing the Elasticsearch `Publish` helpers to make the logic more clear. The new layout makes it much easier to test the error handling and metrics reporting.

The bugs fixed by this refactor are:
- When a batch was split, the events in it were not reported to the observer via `RetryableErrors`. This caused `activeEvents` to increase permanently even after the events were handled.
- When a previously-failed event was ingested to the dead letter index as a raw string, it was reported to the observer as `Acked` (success). The new logic creates a new `dead_letter` metric specifically for this case.
- When a previously-failed event encountered a fatal error ingesting to the dead letter index:
  * It was reported to the observer as both a permanent error and a retryable error, giving incorrect event counts. It's now reported as just a permanent error.
  * It was added to the retry list anyway, which would block all ingestion if the error really was fatal since the queue could never advance past that event. It's now dropped, the same as with a permanent error in the main index.
- If the Elasticsearch bulk index response was invalid, the associated events were dropped and reported as acknowledged

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] ~~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Related issues

- Fixes https://github.com/elastic/beats/issues/39146.

### CODE REVIEW

File: libbeat/outputs/elasticsearch/client.go [hunk 1]
```diff
@@ -380,80 +399,108 @@ func getPipeline(event *beat.Event, defaultSelector *outil.Selector) (string, er
...
 		}
+		return false // ok value
+	}
 
-		client.log.Debugf("Bulk item insert failed (i=%v, status=%v): %s", i, status, msg)
-		stats.fails++
-		failed = append(failed, data[i])
+	if itemStatus == 409 {
+		// 409 is used to indicate an event with same ID already exists if
+		// `create` op_type is used.
```
Comment (cmacknz) [inline libbeat/outputs/elasticsearch/client.go:464]:
```suggestion
		// 409 is used to indicate an event with same ID already exists if
		// `create` op_type is used.
```

You now also get a 409 if the TSDB dimensions are the same for two events, they form an implicit ID. This is probably the more common reason to hit this now.

File: libbeat/outputs/elasticsearch/client.go [hunk 2]
```diff
@@ -76,15 +81,38 @@ type clientSettings struct {
...
+
+	// The API response from Elasticsearch.
+	response eslegclient.BulkResponse
+}
+
 const (
 	defaultEventType = "doc"
 )
 
+var bulkRequestParams = map[string]string{
```
Comment (fearful-symmetry) [inline libbeat/outputs/elasticsearch/client.go:112]: Can we get some comments explaining this ES param? It's a bit odd without any context

Comment (faec) [reply libbeat/outputs/elasticsearch/client.go:114]: Done

File: libbeat/outputs/elasticsearch/client.go (hunk 1 above)
Comment (faec) [reply libbeat/outputs/elasticsearch/client.go]: Added

**Review by belimawr — APPROVED** [review]

File: libbeat/publisher/pipeline/monitoring.go [hunk 3]
```diff
@@ -85,19 +96,46 @@ func newMetricsObserver(metrics *monitoring.Registry) *metricsObserver {
...
+			// queue will accept, or 0 if there is none.
 			queueMaxEvents: monitoring.NewUint(reg, "queue.max_events"),
 
-			activeEvents:     monitoring.NewUint(reg, "events.active"), // Gauge
+			// queue.acked counts events that have been acknowledged by the output
+			// workers. This includes events that were dropped for fatal errors,
+			// which are also reported in events.dropped.
+			queueACKed: monitoring.NewUint(reg, "queue.acked"),
+
+			// (Gauge) queue.filled.pct.events measures the fraction (from 0 to 1)
```
Comment (cmacknz) [inline libbeat/publisher/pipeline/monitoring.go:137]: Can we drop the `.events` from this, ahead of adding support for the queue size in bytes?

People will want to alert on this, they shouldn't have to define it twice.

Comment (faec) [reply libbeat/publisher/pipeline/monitoring.go:137]: Already part of https://github.com/elastic/beats/pull/39774

**Pull request merged**